File: GO/21.merge-two-sorted-lists.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 */
package main

// @lc code=start
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	mergedList := &ListNode{}
	ptr := mergedList

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			ptr.Next = list1
			list1 = list1.Next
		} else {
			ptr.Next = list2
			list2 = list2.Next
		}
		ptr = ptr.Next
	}

	if list1 != nil {
		ptr.Next = list1
	} else {
		ptr.Next = list2
	}

	return mergedList.Next
}

// @lc code=end

File: JS/21.merge-two-sorted-lists.js
/*
 * @lc app=leetcode id=21 lang=javascript
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function (list1, list2) {
  let mergedList = { val: null, next: null };
  let pointer = mergedList;

  while (list1 && list2) {
    // ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
    if (list1.val < list2.val) {
      pointer.next = list1;
      list1 = list1.next;
    } else {
      pointer.next = list2;
      list2 = list2.next;
    }
    // ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
    pointer = pointer.next;
  }

  pointer.next = list1 || list2;

  return mergedList.next;
};
// @lc code=end
